### ISSUE TITLE
Downcasting callable class shouldn't wrap them in an invisible object

### ISSUE THREAD

**Issue by rrousselGit** [dart-lang/language#668]
Consider this callable class:

```dart
class Test {
  void call() {}
}
```

Then it seems that downcasting `Test` to `void Function()` wraps the class in an invisible closure.

This cause undesired side effects such as:
- `runtimeType` is not preserved
- `operator==`/`toString` are impossible to override
- having a `const` constructor on callable classes is ineffective since a new object is created on every downcast anyway.

This behavior is both [undocumented](https://dart.dev/guides/language/language-tour#callable-classes) and unexpected.

Is it because `Test` technically doesn't implement `void Function()`?

Would allowing:
```dart
class Test implements void Function() {
  void call() {
  }
}
```
potentially solve the issue?

**Comment by leafpetersen** [comment]
This was a deliberate change as part of the Dart 2 transition: casting from a class type to a function type is treated as a tear-off of the `.call` method.  This simplifies the underlying representation, since function types are no longer inhabited by both functions and objects.

On the question of tear-offs from const objects, this is something that I think we could in principle allow to be const, which would address your last point.

**Comment by mraleph** [comment]
Link to the original [Breaking Change](https://groups.google.com/a/dartlang.org/d/msg/misc/hvt6jpiYxsE/ZQsrT67ZAQAJ) announcement.

From the implementations perspective this lead to significant simplification of closure calls, making them smaller and more efficient. 

I have filed https://github.com/dart-lang/site-www/issues/2071 to clarify the language tour.

**Comment by rrousselGit** [comment]
Does this mean we can't override ==/toString anymore?

**Comment by mraleph** [comment]
@rrousselGit you can override == and toString - it's just that your object is no longer considered a subtype of a function and can't be appear where function is expected. Meaning that your equality and toString would not be used in those contexts.

Could you outline the use case you have in mind?

**Comment by rrousselGit** [comment]
Nothing worthy of a language modification.

I was just trying something unusual, and got surprised by this undocumented behavior — hence the documentation request. 

TL;DR I used a function as a class, and extension member to emulate the properties.

**Comment by mraleph** [comment]
I am going ahead and close this issue then.

**Comment by rrousselGit** [comment]
As a comeback on this, I've been thinking for a while:
This restricts quite a lot the usefulness of callable classes.

This prevents us from writing:

```dart
void Function() callback;
if (callback is MyCallableClass) {
  // do something more advanced
}
```


Right now, because of this issue, we have to choose between:
- a simple closure for a simple API
- a complex delegate class for more advanced use-cases
